### PR TITLE
Introduce Emformer

### PR DESCRIPTION
**Pull request by hwangjeff**
Adds an implementation of Emformer, a memory-efficient transformer architecture introduced in https://ieeexplore.ieee.org/document/9414560 that targets low-latency streaming speech recognition applications.

### CODE REVIEW

File: torchaudio/prototype/emformer.py [hunk 1]
```diff
@@ -0,0 +1,908 @@
...
+        summary: torch.Tensor,
+        mems: torch.Tensor,
+        left_context_key: torch.Tensor,
+        left_context_val: torch.Tensor,
+    ):
+        r"""Forward pass for inference.
+
+        B: batch size.
+        D: feature dimension of each frame.
+        U: number of utterance frames.
```
Comment (nateanl) [inline torchaudio/prototype/emformer.py:309]: ditto

File: torchaudio/prototype/emformer.py [hunk 2]
```diff
@@ -0,0 +1,908 @@
...
+        dropout (float, optional): dropout probability.
+        activation (str, optional): activation function to use in each Emformer layer's
+            feedforward network.
+        left_context_length (int, optional): length of left context.
+        right_context_length (int, optional): length of right context.
+        segment_length (int, optional): length of each input segment.
+        max_memory_size (int, optional): maximum number of memory elements to use.
+        weight_init_scale_strategy (str, optional): per-layer weight initialization scaling strategy.
+        tanh_on_mem (bool, optional): if ``True``, applies tanh to memory elements.
+        negative_inf (float, optional): value to use for negative infinity in attention weights.
```
Comment (nateanl) [inline torchaudio/prototype/emformer.py:658]: Add default values to the arguments which have it.

**Review by nateanl — COMMENTED** [review]
Some nit suggestions. I need to read the paper before reviewing the implementation details...😅

File: torchaudio/prototype/emformer.py [hunk 3]
```diff
@@ -0,0 +1,908 @@
...
+
+
+__all__ = ["Emformer"]
+
+
+def _lengths_to_padding_mask(
+    lengths: torch.Tensor, max_length: Optional[int] = None
+) -> torch.Tensor:
+    batch_size = lengths.shape[0]
+    if max_length is None:
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply torchaudio/prototype/emformer.py]: yeah it's used only there, and `max_length` isn't used — removed it

File: test/torchaudio_unittest/prototype/emformer_test_impl.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        input, lengths = self._gen_inputs(
+            input_dim, batch_size, num_frames, right_context_length
+        )
+
+        output, output_lengths = emformer(input, lengths)
+
+        self.assertEqual(
+            (batch_size, num_frames - right_context_length, input_dim), output.shape
+        )
+        self.assertEqual((batch_size,), output_lengths.shape)
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply test/torchaudio_unittest/prototype/emformer_test_impl.py:88]: this should suffice — the test generates `lengths` with shape `(batch_size,)`

**Comment by hwangjeff** [comment]
address comments

File: test/torchaudio_unittest/prototype/emformer_test_impl.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hwangjeff) [reply test/torchaudio_unittest/prototype/emformer_test_impl.py:88]: yeah we can add a separate test

**Comment by hwangjeff** [comment]
update tests; fix docs

**Comment by hwangjeff** [comment]
more docstring cleanup

**Review by mthrok — APPROVED** [review]

File: test/torchaudio_unittest/prototype/emformer_test_impl.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        ).to(device=self.device, dtype=self.dtype)
+        return emformer
+
+    def _gen_inputs(self, input_dim, batch_size, num_frames, right_context_length):
+        input = torch.rand(batch_size, num_frames, input_dim).to(
+            device=self.device, dtype=self.dtype
+        )
+        lengths = torch.randint(1, num_frames - right_context_length, (batch_size,)).to(
+            device=self.device, dtype=self.dtype
+        )
```
Comment (csukuangfj) [inline test/torchaudio_unittest/prototype/emformer_test_impl.py:26]: You have to add something like below to ensure that `max(lengths)` equals to `inpute.size(1) - right_context_length`.
```python3
lenghts[0] = num_frames - right_context_length
```

**Pull request merged**